repositories/src/main/scala/cromwell/pipeline/datastorage/dto/User.scala

```diff
@@ -1,35 +1,76 @@
 package cromwell.pipeline.datastorage.dto
 
-import cromwell.pipeline.datastorage.dto.User.UserEmail
-import play.api.libs.functional.syntax._
-import play.api.libs.json.Reads._
+import cats.data.{ NonEmptyChain, Validated }
+import cromwell.pipeline.utils.validator.Wrapped
 import play.api.libs.json.{ Format, Json, OFormat }
 import slick.lifted.MappedTo
 
 final case class User(
-  userId: UserId,
+  userId: UUID,
   email: UserEmail,
   passwordHash: String,
   passwordSalt: String,
-  firstName: String,
-  lastName: String,
+  firstName: Name,
+  lastName: Name,
   profilePicture: Option[ProfilePicture] = None,
   active: Boolean = true
 )
 
 object User {
   implicit lazy val userFormat: OFormat[User] = Json.format[User]
-  type UserEmail = String
-}
-
-final case class UserId(value: String) extends MappedTo[String]
-
-object UserId {
-  implicit lazy val userIdFormat: Format[UserId] = implicitly[Format[String]].inmap(UserId.apply, _.value)
 }
 
 final case class ProfilePicture(value: Array[Byte]) extends MappedTo[Array[Byte]]
 
 object ProfilePicture {
   implicit lazy val profilePictureFormat: OFormat[ProfilePicture] = Json.format[ProfilePicture]
 }
+
+final class UserEmail private (override val unwrap: String) extends AnyVal with Wrapped[String]
+
+object UserEmail extends Wrapped.Companion {
+  type Type = String
+  type Wrapper = UserEmail
+  type Error = String
+  implicit lazy val emailFormat: Format[UserEmail] = Json.valueFormat[UserEmail]
+  override protected def create(value: String): UserEmail = new UserEmail(value)
+  override protected def validate(value: String): ValidationResult[String] = Validated.cond(
+    value.matches("^[^@]+@[^\\.]+\\..+$"),
```

Comment:
Is text1@text2@.text3@ valid email?

Reply:
I think there is no perfect regex for email, best way to approve email is sending to users an activation link imho, but if you have this one share please